src/Storages/StorageMongoDB.cpp

```diff
@@ -28,6 +35,49 @@ namespace ErrorCodes
 {
     extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
     extern const int MONGODB_CANNOT_AUTHENTICATE;
+    extern const int MONGODB_ERROR;
+    extern const int EMPTY_DATA_PASSED;
+}
+
+namespace
+{
+    using MongoArray = Poco::MongoDB::Array;
+
+    DataTypePtr getDataTypeMongoDB(const Poco::MongoDB::Element & value)
+    {
+        switch (value.type())
+        {
+            case Poco::MongoDB::ElementTraits<Poco::MongoDB::NullValue>::TypeId:
+                return std::make_shared<DataTypeNothing>();
+            case Poco::MongoDB::ElementTraits<bool>::TypeId:
+                return std::make_shared<DataTypeUInt8>();
+            case Poco::MongoDB::ElementTraits<Int32>::TypeId:
+                return std::make_shared<DataTypeInt32>();
+            case Poco::MongoDB::ElementTraits<Poco::Int64>::TypeId:
+                return std::make_shared<DataTypeInt64>();
+            case Poco::MongoDB::ElementTraits<Float64>::TypeId:
+                return std::make_shared<DataTypeFloat64>();
+            case Poco::MongoDB::ElementTraits<Poco::Timestamp>::TypeId:
+                return std::make_shared<DataTypeDateTime>();
+            case Poco::MongoDB::ElementTraits<MongoArray::Ptr>::TypeId:
+            {
+                auto parent = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+
+                DataTypes types;
+                types.reserve(parent->size());
+
+                for (size_t idx = 0, size = parent->size(); idx < size; ++idx)
+                {
+                    Poco::MongoDB::Element::Ptr child = parent->get(static_cast<int>(idx));
+                    types.push_back(getDataTypeMongoDB(*child));
+                }
+                return std::make_shared<DataTypeTuple>(types);
+            }
+            default:
+                return std::make_shared<DataTypeString>();
```

Comment:
Why? We don't support reading any type as String from Mongo (maybe also good feature for later improvements). I think we should throw an exception that type is not supported. Also we support one more Mongo type: `ObjectId` as our `String`.